### PR TITLE
Replace @fabs builtin with new @abs builtin

### LINKED ISSUE #16026: Proposal: rename @fabs to @abs and accept integers and vectors of integers too
This proposal is to merge `std.math.absCast`, `std.math.absInt`, and `@fabs` that currently only accepts floats and vectors of floats into an `@abs` and make that `@abs` additionally accept integers too as well as vectors of integers.

This new `@abs` converts and returns an unsigned integer if the argument is a signed integer. Same for vectors.

Status quo: `@fabs`, [`std.math.absCast`](https://github.com/ziglang/zig/blob/38fc826a5a82bbe11a893be11d3cc1439c477ee8/lib/std/math.zig#L1101C1-L1136), or [`std.math.absInt`](https://github.com/ziglang/zig/blob/38fc826a5a82bbe11a893be11d3cc1439c477ee8/lib/std/math.zig#L867C1-L910) depending on your specific case.

* `absCast` accepts signed or unsigned and always returns an unsigned integer so if the input is signed, it converts to unsigned.
* `absInt` only takes a signed integer and returns a signed integer.

These are clunky and it takes a while to understand their differences.

With `@abs`, however, we always return an unsigned integer or a vector of unsigned integers if an integer or a vector of integers was given and the reason is that, generally, the user is perfectly fine with an unsigned integer type: when I grep the Zig codebase for `absInt` and look how `absInt` is used in those specific cases, there are actually **zero legitimate real-world uses of `absInt`** (at least in the Zig codebase).
If you look for those `absInt` occurrences yourself you'll find that it's used either in tests or to implement some obscure C function (`__builtin_abs`).

Compare that to the **13** real-world uses of `absCast` that I found in the Zig codebase. So, you really just want `@fabs` or `absCast`.

So, new status quo: if you need a signed integer, just convert it yourself.

The advantage of combining all three into one is that you don't have to be confused about why we don't have `std.math.abs` or `@abs`. You no longer have to think about which one to use.

This was originally brought up by @jacobly0 here: https://github.com/ziglang/zig/pull/15879#discussion_r1220561951

## Performance

There is actually also a performance benefit to be had here. As @jacobly0 pointed out to me, x86 actually has instructions for vectors of integers.

Currently, none of the 3 options can really take advantage of that in such a way those hardware instructions are emitted for sure. With an `@abs` builtin we can do that.
`absInt` accepts vectors and uses vector builtins but still we have to rely on the optimizer there.

Finally, this should allow the backends to emit better code for regular integers too.
For example in LLVM we don't seem to be using `llvm.abs.*`.

### PR DESCRIPTION
**Pull request by antlilja**
Closes #16026

This PR replaces the `@fabs` builtin with a new `@abs` builtin accepts signed integers in addition to floating point types.

I've implemented codegen for the llvm, x86_64, wasm and c backends (no vector integer abs except on llvm). 

As I've implemented it right now the builtin does not accept unsigned integers only signed integers and always outputs unsigned integers. My rationale was that an unsigned integer is never negative and thus `@abs` isn't a valid operation. The other option would be to just nop for `@abs` on an unsigned value.

~~I've also not implemented any safety yet in case of overflow.~~

### CODE REVIEW

**Comment by mgord9518** [comment]
I think not accepting unsigned ints makes more sense than making it a noop

**Comment by matu3ba** [comment]
> I've also not implemented any safety yet in case of overflow.

IEEE754 floats can not overflow and I dont see `@abs(-inf), @abs(inf), @abs(-0.0), @abs(+0.0), @abs(-Nan), @abs(Nan), @abs(-denormal), @abs(denormal)`. LLVM does not preserve quite and signaling NaN (it uses any NaN) and this is undocumented, so I would not trust it too much regarding float correctness.

**Comment by matu3ba** [comment]
CI fails with
```
[16/21] Running zig1.wasm to produce /home/ci/actions-runner4/_work/zig/zig/build-debug/zig2.c
FAILED: zig2.c 
cd /home/ci/actions-runner4/_work/zig/zig && /home/ci/actions-runner4/_work/zig/zig/build-debug/zig1 /home/ci/actions-runner4/_work/zig/zig/lib build-exe src/main.zig -ofmt=c -lc -OReleaseSmall --name zig2 -femit-bin="/home/ci/actions-runner4/_work/zig/zig/build-debug/zig2.c" --mod build_options::/home/ci/actions-runner4/_work/zig/zig/build-debug/config.zig --deps build_options -target aarch64-linux
src/codegen/llvm.zig:8449:93: error: member function expected 5 argument(s), found 7
src/codegen/llvm/bindings.zig:655:12: note: function declared here
ninja: build stopped: subcommand failed.
```

bindings points to
```
    pub const buildCall = LLVMBuildCall2;
    extern fn LLVMBuildCall2(
        *Builder,
        *Type,
        Fn: *Value,
        Args: [*]const *Value,
        NumArgs: c_uint,
        Name: [*:0]const u8,
    ) *Value;
```
which has been changed in https://github.com/ziglang/zig/commit/d21d1d4ba25701410b9ae99921326f8bfc8a44dd.

**Comment by antlilja** [comment]
I seem to be running into the same LLVM bug on x86_64 macos reported here: #12827

**Comment by antlilja** [comment]
The failing test has been skipped on x86_64-macos because it was due to a bug in LLVM

**Comment by antlilja** [comment]
I've now also implemented the ```llvm.abs.``` intrinsic for the new LLVM builder API.

**Review by andrewrk — CHANGES_REQUESTED** [review]
Thanks, nice work.

Before this is merged, it needs `zig fmt` to auto-convert `@fabs` to `@abs`.

Also, the absolute value function is well-defined for positive integers. The absolute value of a positive integer is the same positive integer. Just because `u8` cannot store the value `-10`, does not mean that `abs(10) != 10`.

File: lib/std/math.zig [hunk 1]
```diff
@@ -1003,46 +958,9 @@ fn testRem() !void {
 
 /// Returns the absolute value of a floating point number.
 /// Uses a dedicated hardware instruction when available.
-/// This is the same as calling the builtin @fabs
+/// This is the same as calling the builtin @abs
 pub inline fn fabs(value: anytype) @TypeOf(value) {
```
Comment (wooster0) [inline lib/std/math.zig:962]:
```suggestion
pub inline fn abs(value: anytype) @TypeOf(value) {
```

File: doc/langref.html.in [hunk 2]
```diff
@@ -9424,14 +9424,14 @@ fn doTheTest() !void {
       Supports {#link|Floats#} and {#link|Vectors#} of floats.
       </p>
       {#header_close#}
-      {#header_open|@fabs#}
-      <pre>{#syntax#}@fabs(value: anytype) @TypeOf(value){#endsyntax#}</pre>
+      {#header_open|@abs#}
+      <pre>{#syntax#}@abs(value: anytype) @TypeOf(value){#endsyntax#}</pre>
       <p>
-      Returns the absolute value of a floating point number. Uses a dedicated hardware instruction
+      Returns the absolute value of a floating point or signed integer number. Uses a dedicated hardware instruction
```
Comment (wooster0) [inline doc/langref.html.in:9430]: Looking at `Sema`, unsigned integers seem to be supported.
```suggestion
      Returns the absolute value of an integer or a floating point number. Uses a dedicated hardware instruction
```

File: lib/std/math.zig (hunk 1 above)
Comment (wooster0) [inline lib/std/math.zig:962]:
```suggestion
pub inline fn abs(value: anytype) @TypeOf(value) {
```

**Comment by andrewrk** [comment]
This looked like it was pretty close to being landed, however, it did not get to green CI status (perhaps due to a flaky test in master branch that needs to get disabled?), and it has a few conflicts to be resolved. Since it has not been touched in over 1 month, I'm going to close it. However, please feel free to open a new pull request if you wish to continue these efforts.

**Pull request closed without merging**